### PR TITLE
feat(assert/unstable): add `assertNever`

### PR DESCRIPTION
**Pull request by magurotuna**
This commit adds a new function `assertNever` to the `assert` submodule.

This function is particularly useful when we want to check whether we properly handle all the variants of a discriminated union.
Ref: https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#union-exhaustiveness-checking

### CODE REVIEW

File: assert/never.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * Make an assertion that `x` is of type `never`.
```
Comment (kt3k) [inline assert/never.ts:30]: We now have a policy of accepting all new features as unstable API. Could you add `**UNSTABLE**: New API, yet to be vetted.` note at the beginning of jsdoc and also add `@experimental` jsdoc tag? https://github.com/denoland/std/blob/010784cd03d0b3ac7e913d82d3468d293adc13c8/.github/CONTRIBUTING.md#notices-for-unstable-apis

File: assert/never.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * If not then throw.
+ *
+ * > [!WARNING]
+ * > **UNSTABLE**: New API, yet to be vetted.
```
Comment (kt3k) [inline assert/never.ts:34]: I'll update the format a bit. (We're removing markdown `!WARNING` directive as it's not rendered well in VSCode https://github.com/denoland/std/pull/5672)

```suggestion
 * **UNSTABLE**: New API, yet to be vetted.
 *
 * Make an assertion that `x` is of type `never`.
 * If not then throw.
```

**Review by kt3k — APPROVED** [review]
LGTM

**Comment by magurotuna** [comment]
@kt3k 
Minor question, should we fix [the instruction](https://github.com/denoland/std/blob/8d1d2d773990b875a7d7c4f7a4516d3e3051b47a/.github/CONTRIBUTING.md#notices-for-unstable-apis) I referred to so we'll place `<description>` after the UNSTABLE warning and remove `[!WARNING]`?

**Comment by kt3k** [comment]
I think so, but @iuioiua might have a different idea. ref https://github.com/denoland/std/pull/5672

**Comment by magurotuna** [comment]
Got it. I'll watch that issue 👍

**Comment by iuioiua** [comment]
This is the same as [`unreachable()`](https://jsr.io/@std/assert/doc/unreachable/~/unreachable).

**Comment by magurotuna** [comment]
@iuioiua I thought that way, but actually `unreachable` does not do any compile-time checking while `assertNever` does. 

This playground illustrates this:
https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABAQwM6oKYCcoDkMBu2AFAB4BciYh2AlJdUVogN4BQiiUAFlnAO5UMggKJY+WYgCJG2KbQDcbAL5s2oSLASJwWDMgjdkAIwA2GYgFtUAc0qooWGGBv0hTVhy68BQ0eLhJAANdfUMTc0QAEhZrG2UgxRU1KABPAAcMRAAVZBtEAF5EKQBBKUQAH2KAISkldXBoeCQQTBL0bDwaSSg8ylzXT05UfhgoQ0RiXsH2Tk4INCzSqXIvOcQAeg3EABM4RFQ4SwweZxs1ueM9ZABrJXWdjGBkEFMoVfXOLcRgQJs4KBcfZGMA7SJSWoAOhy3BgqEQelQryg8OcKC4GSy2AkkIunDQmBw+CYUzySU4qlUDU0zR0mAAqmBruEzBZpv0yUMDqNxtxJtNaFz5otimUPp9vnsDkcTrCXBdHs9keL1t9flh-oCoMDkKDwVCYXDdnAMPCIEd0jBzAAaFCg7x8fjw5CArDgWDHXGfHRMsJGVnSOBwdKoeT3RCUoA

**Pull request merged**